accounting-service/app/src/main/java/com/microservices/accounting/service/PaymentServiceImpl.java

```diff
@@ -0,0 +1,84 @@
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import javax.servlet.UnavailableException;
+import java.util.Objects;
+import java.util.Random;
+
+@Service
+public class PaymentServiceImpl implements PaymentService {
+    private PaymentRepository paymentRepository;
+    private final Logger logger = LoggerFactory.getLogger(PaymentServiceImpl.class);
+
+    @Autowired
+    public PaymentServiceImpl(PaymentRepository paymentRepository) {
+        this.paymentRepository = paymentRepository;
+    }
+
+    @Override
+    public PaymentDetailsDto invokePayment(InvokePaymentDto invokePaymentDto) {
+        Objects.requireNonNull(invokePaymentDto);
+
+        logger.info("Invoking payment: {}...", invokePaymentDto);
+        CardInfo cardInfo = invokePaymentDto.getUser().getCardInfo();
+        PaymentStatus paymentStatus = (cardInfo == CardInfo.VALID && isRequestApproved())
+                ? PaymentStatus.ACCEPTED
+                : PaymentStatus.DENIED;
+
+        PaymentEntity savedPayment = paymentRepository.save(new PaymentEntity(invokePaymentDto, paymentStatus));
+        logger.info("Saved payment: {}", savedPayment);
+        return new PaymentDetailsDto(savedPayment);
+    }
+
+    @Override
+    @Transactional
+    public PaymentDetailsDto revertPayment(RevertPaymentDto revertPaymentDto) throws UnavailableException {
+        Objects.requireNonNull(revertPaymentDto);
+        logger.info("Reverting payment: {}", revertPaymentDto);
+
+        PaymentEntity paymentToRevert = paymentRepository.findById(revertPaymentDto.getPaymentId())
+                .orElseThrow(() -> new IllegalArgumentException("No payment with id " + revertPaymentDto.getPaymentId()));
+
+        switch (paymentToRevert.getPaymentStatus()) {
+            case REVERTED:
+                logger.info("Payment with id {} is already reverted", paymentToRevert.getPaymentId());
+                return new PaymentDetailsDto(paymentToRevert);
+            case DENIED:
+                logger.error("Payment with id {} cannot be reverted. Has a denied status", paymentToRevert.getPaymentId());
+                throw new IllegalArgumentException("Denied payment cannot be reverted");
+            case ACCEPTED:
+                if (isRequestApproved()) {
+                    paymentToRevert.setPaymentStatus(PaymentStatus.REVERTED);
+                    PaymentEntity updatedPaymentEntity = paymentRepository.save(paymentToRevert);
+                    logger.info("Payment with id {} is reverted");
+                    return new PaymentDetailsDto(updatedPaymentEntity);
+                } else {
+                    logger.error("Payment reversion cannot be processed");
+                    throw new UnavailableException("Payment reversion cannot be processed now. Try later");
+                }
+            default:
+                throw new AssertionError("Unknown payment status");
```

Comment:
Assertion?

Reply:
As far as I know, this is a common used exception for enum switch. In case of adding new enum value, you would get an exception

Reply:
IllegalArgumentException

Reply:
This is an exception that is used in "Effective java". We throw an IllegalArgument to indicate that client sent a wrong request. So client could handle it and try with another param (another id). This is not a client's mistake. 

IllegalArgument and AssertionError are both used in switch, but, as I said, in that case I would prefer not to use IllegalArg